rook/rook
# ceph: Adding support for admission controllers

**Description of your changes:**
Adds various scripts and code changes to deploy admission controllers alongside rook. 

**What are admission controllers**
Admission controllers help in validating the CRDs at native kubernetes  level, accepting or rejecting an api request based on our custom criteria

For example, The PR here attempts to provide solution for one of the issues mentioned in  https://github.com/rook/rook/issues/4819

**Cannot change the dataDirHostPath after initial creation**

Our webhooks have the logic defined in them to to continuously check if an api request is an update for an existing ceph cluster and validates if the dataDirHostPath has been changed or not. 

If it has been changed from initial creation, it will reject the request and throw the below error

`for: "cluster.yaml": admission webhook "webhook-server.rook-ceph.svc" denied the request: Invalid Update : DataDirHostPath change from /var/lib/rook to /var/rook is not allowed
`


More such validations can be applied for any kind of resource

**How to test this PR** 
1. Clone
2. Create any cluster
3. Run the following commands
```
kubectl create -f common.yaml
./config-admission-controllers.sh
```
4. Now that the secrets have been deployed, we can deploy the operator and create a cluster using the following commands
```
kubectl create -f operator.yaml
kubectl create -f cluster.yaml
```
5. The operator on detecting the secrets will deploy the admission controllers and start listening for all validations
6. Make changes to cluster.yaml in spec.datadirhostpath 
  `kubectl apply -f cluster.yaml`

On doing all the above, we could see the request being rejected.

**Which issue is resolved by this Pull Request:**
https://github.com/rook/rook/issues/2363
https://github.com/rook/rook/issues/4819

[test ceph]
